docs: Describe cloudformation.yaml for docs (#4394)
aws/karpenter: website/content/en/preview/concepts/setup.md

Reviewer: Overall, we / you could move much of the item-by-item explanation into comments within the template itself. Then this new page becomes more of a context setting explanation that segues into a deployment guide.

Author: This is a good thought. I'd like to merge this PR and consider this after.

      # The PolicyDocument must be in JSON string format because we use a StringEquals condition that uses an interpolated
      # value in one of its key parameters which isn't natively supported by CloudFormation
      PolicyDocument: !Sub |
        {
          "Version": "2012-10-17",
          "Statement": [
```

### AllowScopedEC2InstanceActions

The AllowScopedEC2InstanceActions statement ID (Sid) identifies a set of EC2 resources that are allowed to be used with
[RunInstances](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_RunInstances.html) and [CreateFleet](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateFleet.html) actions.

```
            {
              "Sid": "AllowScopedEC2InstanceActions",
              "Effect": "Allow",
              "Resource": [
                "arn:${AWS::Partition}:ec2:${AWS::Region}::image/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}::snapshot/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:spot-instances-request/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:security-group/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:subnet/*",

>                 "arn:${AWS::Partition}:ec2:${AWS::Region}:*:launch-template/*"

Reviewer: How come this isn't locked down further (eg using tags and tag conditions)? It overrides a more specific statement later in the template.

Author: I'm not sure. I'll pass that information on to the engineering team.

Reviewer 2: AFAIK, there isn't a way to lock down calls to `RunInstances` and `CreateFleet` for these resources further without knowing more information about those resources themselves.

For instance, `ec2:RunInstances` will use an image in your account as part of the instance launch. Getting access to that instance requires permission on that image, but Karpenter wasn't involved in creating that image and, therefore, has no specific knowledge about what the properties of that image are at all. A proper thing to do for most users would be to lock down the resources that Karpenter can interact with here by tagging these resources in a custom way and then using a `ResourceTag` condition to ensure that Karpenter only uses the resources that it has specific access to use

Reviewer: > A proper thing to do for most users would be to lock down the resources that Karpenter can interact with here by tagging these resources in a custom way and then using a ResourceTag condition to ensure that Karpenter only uses the resources that it has specific access to use.

Yep, I usually do this. You can also restrict by VPC, by the tag that the launch template has, and more.

An AWS security specialist can explain more about the options here (also, @jonathan-innis, feel free to message me on Kubernetes Slack to follow up directly).

Author: I have updated to the latest cloudformation.yaml file to be used for v1beta1 since these comments entered. Please check to see if the updated content satisfies these concenrs.

```
            {
              "Sid": "AllowRegionalReadActions",
              "Effect": "Allow",
              "Resource": "*",
              "Action": [
                "ec2:DescribeAvailabilityZones",
                "ec2:DescribeImages",
                "ec2:DescribeInstances",
                "ec2:DescribeInstanceTypeOfferings",
                "ec2:DescribeInstanceTypes",
                "ec2:DescribeLaunchTemplates",
                "ec2:DescribeSecurityGroups",
                "ec2:DescribeSpotPriceHistory",
                "ec2:DescribeSubnets"
              ],
              "Condition": {
                "StringEquals": {
                  "aws:RequestedRegion": "${AWS::Region}"
                }
              }
            },
```

### AllowGlobalRead Actions

```
            {
              "Sid": "AllowGlobalReadActions",
              "Effect": "Allow",
              "Resource": "*",
              "Action": [
                "pricing:GetProducts",

>                 "ssm:GetParameter"

Reviewer 2: This has been restricted since the writing of this doc. (see the current `cloudformation.yaml` file that's currently in the preview docs)

---
title: "Karpenter CloudFormation Setup"
linkTitle: "CloudFormation Setup"
weight: 5
description: >
  Descriptions of how Karpenter uses CloudFormation to set up permissions
---

> When you create a cluster to use with Karpenter in the [Getting Started with Karpenter]({{< relref "../getting-started/getting-started-with-karpenter" >}}) guide, the procedure uses CloudFormation to prepare the cluster for Karpenter to be able to create and manage nodes, as well as gather and respond to interruption events.

Author: Done.

This document describes the `cloudformation.yaml` file used in that guide.

> These descriptions will be useful to understand:

Author: Done

> * What Karpenter is authorized to do with your EKS cluster and AWS resources when you use the `cloudformation.yaml` file

Author: Done.

* What permissions you need to set up if you are adding Karpenter to an existing cluster

# Review the cloudformation.yaml file

To download a particular version of `cloudformation.yaml`, set the version and use `curl` to pull the file to your local system:

```bash

> export KARPENTER_VERSION=v0.29.0

Author: Done.

curl https://raw.githubusercontent.com/aws/karpenter/"${KARPENTER_VERSION}"/website/content/en/preview/getting-started/getting-started-with-karpenter/cloudformation.yaml > cloudformation.yaml
```

> The `cloudformation.yaml` file starts with the following information:

Author: Done.

```
AWSTemplateFormatVersion: "2010-09-09"
Description: Resources used by https://github.com/aws/karpenter
Parameters:
  ClusterName:
    Type: String
    Description: "EKS cluster name"
Resources:
```

The rest of the `cloudformation.yaml` file describes the resources that CloudFormation deploys.
The sections of that file can be grouped together under the following general headings:

> * **Node Authorization**: Creates a NodeInstanceProfile, attaches a NodeRole to it, and connects it to an IAM Identity Mapping that Karpenter uses. This defines the permissions each node managed by Karpenter has to access EC2 and other AWS resources.

Author: Done.

> * **Karpenter Controller Authorization**:  Creates a service account (named `karpenter`) that is combined with the KarpenterControllerPolicy. The KarpenterControllerPolicy allows the Karpenter controller to control assets within EC2 specifically and within AWS in general.

Author: Done.

> * **Interruption Handling**: The interruption handling sections of this file allow the Karpenter controller to see and respond to interruptions that occur with the nodes that Karpenter is managing. Interruptions can reflect things like a spot instance going away or physical hardware going down. Allowing the Karpenter controller to see these interruptions allows Karpenter to respond by bringing nodes up and down and moving workloads.

Author: Done.

A lot of the object naming that is done by `cloudformation.yaml` is based on the following:

> * Cluster name: With a user name of `joe` the Getting Started Guide would name your cluster `joe-karpenter-demo`

Reviewer 2: ```suggestion
* Cluster name: With a user name of `joe` the Getting Started Guide would name your cluster `joe-karpenter-demo`
```

Just curious: Is there a typical name that AWS docs typically use as generic usernames? Some good examples for common names in the security space (for example) are "Bob" and "Alice"

Author: Honestly, I don't know. But bob is fine with me. I'll use that.

Author: Done

That name would then be appended to any name below where `${ClusterName}` is included.

* Partition: Any time an ARN is used, it includes the partition name to identify where the object is found. In most cases, that partition name is `aws`. However, it could also be `aws-cn` (for China Regions) or `aws-us-gov` (for AWS GovCloud US Regions).

> # Node Authorization in cloudformation.yaml

Reviewer 2: ```suggestion
# Node Authorization
```

Author: Done

        - !Sub "arn:${AWS::Partition}:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly"
        - !Sub "arn:${AWS::Partition}:iam::aws:policy/AmazonSSMManagedInstanceCore"
```

The role created here includes several AWS managed policies, which are designed to provide permissions for specific uses needed by the nodes to work with EC2 and other AWS resources. These include:

* [AmazonEKS_CNI_Policy](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonEKS_CNI_Policy.html): Provides the permissions that the Amazon VPC CNI Plugin needs to configure EKS worker nodes.
* [AmazonEKSWorkerNodePolicy](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonEKSWorkerNodePolicy.html): Lets Amazon EKS worker nodes connect to EKS Clusters.
* [AmazonEC2ContainerRegistryReadOnly](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonEC2ContainerRegistryReadOnly.html): Allows read-only access to repositories in the Amazon EC2 Container Registry.
* [AmazonSSMManagedInstanceCore](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonSSMManagedInstanceCore.html): Adds AWS Systems Manager service core functions for Amazon EC2.

> # Karpenter Controller Authorization in cloudformation.yaml

Reviewer 2: ```suggestion
# Karpenter Controller Authorization
```

Author: Done

This section sets the permissions that the Karpenter Controller has to create and manage EC2 and other AWS resources.
In particular, the section creates a service account (karpenter) that is combined with the KarpenterControllerPolicy.
The permissions here go beyond the permissions assigned to Karpenter nodes in the previous section.

The resources defined in this section include:

* KarpenterControllerPolicy

> * KarpenterInterruptionQueue

Reviewer 2: This and the other things down to L142 don't apply to the controller authorization

Author: Thanks. I forgot to fix that. Done.

The `iam:PassedToService` restricts the permission to make these requests to only the EC2 service (`ec2.amazonaws.com`).

```
            {
              "Sid": "AllowPassingInstanceRole",
              "Effect": "Allow",
              "Resource": "arn:${AWS::Partition}:iam::${AWS::AccountId}:role/KarpenterNodeRole-${ClusterName}",
              "Action": "iam:PassRole",
              "Condition": {
                "StringEquals": {
                  "iam:PassedToService": "ec2.amazonaws.com"
                }
              }
            },
```

### AllowAPIServerEndpointDiscovery
The Karpenter controller needs to be able to find the Kubernetes cluster's API endpoint in order to communicate with the API server.
The AllowAPIServerEndpointDiscovery Sid allows the Karpenter controller to get that information (`eks:DescribeCluster`) for the cluster (`cluster/${ClusterName}`).
```
            {
              "Sid": "AllowAPIServerEndpointDiscovery",
              "Effect": "Allow",
              "Resource": "arn:${AWS::Partition}:eks:${AWS::Region}:${AWS::AccountId}:cluster/${ClusterName}",
              "Action": "eks:DescribeCluster"
            }
          ]
        }
```

> # Interruption Handling in cloudformation.yaml

Reviewer 2: ```suggestion
# Interruption Handling
```

A lot of the object naming that is done by `cloudformation.yaml` is based on the following:

* Cluster name: With a user name of `bob` the Getting Started Guide would name your cluster `bob-karpenter-demo`
That name would then be appended to any name below where `${ClusterName}` is included.

* Partition: Any time an ARN is used, it includes the partition name to identify where the object is found. In most cases, that partition name is `aws`. However, it could also be `aws-cn` (for China Regions) or `aws-us-gov` (for AWS GovCloud US Regions).

# Node Authorization 

The following sections of the `cloudformation.yaml` file set up permissions related to what Kubernetes nodes created by Karpenter can do with EC2 and other AWS features.
In particular, this involves setting up an instance profile and attaching a node role to that profile with the following objects:

* KarpenterNodeInstanceProfile

> * KarpenterNodeRole

Reviewer: I'd list the role first, then the instance profile that depends on it.

Author: Done.

## KarpenterNodeInstanceProfile
This section creates an [EC2 Instance Profile](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2_instance-profiles.html) that includes the node role named `KarpenterNodeRole`, with the cluster name appended.
For example, with a cluster name of `bob-karpenter-demo`, the instance profile name would look like:

`KarpenterNodeInstanceProfile-bob-karpenter-demo`

```
  KarpenterNodeInstanceProfile:
    Type: "AWS::IAM::InstanceProfile"
    Properties:
      InstanceProfileName: !Sub "KarpenterNodeInstanceProfile-${ClusterName}"
      Path: "/"
      Roles:
        - !Ref "KarpenterNodeRole"
```

> To do this manually for an existing cluster, you would find your cluster's InstanceProfileName and use that to attach the role needed.

Reviewer: ```suggestion
To do this manually for an existing cluster, you would find the name of an instance profile that your
Kubernetes nodes are already using, and use that to associate the existing IAM role to
Karpenter-managed nodes.
```
?

Author: Done.

To list all instance profiles, type:

```bash
aws iam list-instance-profiles
```

To see information for a particular instance profile (for example, `MyRoleForInstances`), type:

```bash
aws iam get-instance-profile --instance-profile-name "MyRoleForInstances"
```
```
INSTANCEPROFILE arn:aws:iam::973227887653:instance-profile/MyRoleForInstances  2020-12-10T19:45:31+00:00       AIPA6FGHHBQS4F2FFHYYV   MyRoleForInstances     /
ROLES   arn:aws:iam::111111111111:role/MyRoleForInstances      2020-12-10T19:45:30+00:00       /       AROA6FGHHBQSQ5NKPJT2H   MyRoleForInstances
ASSUMEROLEPOLICYDOCUMENT        2012-10-17
STATEMENT       sts:AssumeRole  Allow
PRINCIPAL       ec2.amazonaws.com
```

## KarpenterNodeRole

> This section creates the node role that is attached to the `KarpenterNodeInstanceProfile` instance profile created earlier.

Reviewer: ```suggestion
This section of the template defines a node IAM role. Karpenter also needs an instance profile
to associate the IAM role with EC2 instances as they launch.
```

Author: Done.

```

The role created here includes several AWS managed policies, which are designed to provide permissions for specific uses needed by the nodes to work with EC2 and other AWS resources. These include:

* [AmazonEKS_CNI_Policy](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonEKS_CNI_Policy.html): Provides the permissions that the Amazon VPC CNI Plugin needs to configure EKS worker nodes.
* [AmazonEKSWorkerNodePolicy](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonEKSWorkerNodePolicy.html): Lets Amazon EKS worker nodes connect to EKS Clusters.
* [AmazonEC2ContainerRegistryReadOnly](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonEC2ContainerRegistryReadOnly.html): Allows read-only access to repositories in the Amazon EC2 Container Registry.
* [AmazonSSMManagedInstanceCore](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonSSMManagedInstanceCore.html): Adds AWS Systems Manager service core functions for Amazon EC2.

# Karpenter Controller Authorization 

This section sets the permissions that the Karpenter Controller has to create and manage EC2 and other AWS resources.

> In particular, the section creates a service account (karpenter) that is combined with the KarpenterControllerPolicy.

Author: Done.

> The permissions here go beyond the permissions assigned to Karpenter nodes in the previous section.

Author: Done.

The resources defined in this section are associated with:

* KarpenterControllerPolicy

Because the scope of the KarpenterControllerPolicy is an AWS region, the cluster's AWS region is included in the `AllowScopedEC2InstanceActions`.

## KarpenterControllerPolicy

A `KarpenterControllerPolicy` object sets the name of the policy, then defines a set of resources and actions allowed for those resources.

> The policy creates authorization for the Karpenter Controller to AWS resources, instead of using the NodeInstanceProfile which probably has less permissions to AWS resources.

Reviewer: Another option is to combine this with the notes about the role, as the two go together.

> The policy creates authorization for the Karpenter Controller to AWS resources, instead of using the NodeInstanceProfile which probably has less permissions to AWS resources.

Author: I dropped it.

For our example, the KarpenterControllerPolicy would be named: `KarpenterControllerPolicy-bob-karpenter-demo`

```
  KarpenterControllerPolicy:
    Type: AWS::IAM::ManagedPolicy
    Properties:
      ManagedPolicyName: !Sub "KarpenterControllerPolicy-${ClusterName}"
      # The PolicyDocument must be in JSON string format because we use a StringEquals condition that uses an interpolated
      # value in one of its key parameters which isn't natively supported by CloudFormation
      PolicyDocument: !Sub |
        {
          "Version": "2012-10-17",
          "Statement": [
```
Someone wanting to add Karpenter to an existing cluster, instead of using `cloudformation.yaml`, would need to find a way to create the policy and assign that policy to the service account to use IRSA.

### AllowScopedEC2InstanceActions

The AllowScopedEC2InstanceActions statement ID (Sid) identifies a set of EC2 resources that are allowed to be used with
[RunInstances](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_RunInstances.html) and [CreateFleet](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateFleet.html) actions.

> For `RunInstances` and `CreateFleet` actions, the Karpenter controller can access `image`, `snapshot`, `spot-instances-request`, `security-group`, `subnet` and `launch-template` EC2 resources, scoped for the particular AWS partition and region.

Author: Got it. Done.

action for all Karpenter provisioners, and requires that the `kubernetes.io/cluster/${ClusterName}` tag be set to `owned` and that the `karpenter.sh/provisioner-name` tag be set to any value with these actions.
This makes sure that these Karpenter-managed resources are assigned these tags.

```
            {
              "Sid": "AllowScopedEC2LaunchTemplateActions",
              "Effect": "Allow",
              "Resource": "arn:${AWS::Partition}:ec2:${AWS::Region}:*:launch-template/*",
              "Action": "ec2:CreateLaunchTemplate",
              "Condition": {
                "StringEquals": {
                  "aws:RequestTag/kubernetes.io/cluster/${ClusterName}": "owned"
                },
                "StringLike": {
                  "aws:RequestTag/karpenter.sh/provisioner-name": "*"
                }
              }
            },
```

### AllowScopedEC2InstanceActionsWithTags
The AllowScopedEC2InstanceActionsWithTags Sid allows the 
[RunInstances](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_RunInstances.html) and [CreateFleet](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateFleet.html)

> actions requested by the Karpenter controller to access all `fleet`, `instance`, `volume` or `network-interface` EC2 resources (for the partition and region), and requires that the `kubernetes.io/cluster/${ClusterName}` tag be set to `owned` and a `karpenter.sh/provisioner-name` tag be set to any value with these actions.

Author: Done.

This makes sure that these resources that are managed by the Karpenter controller are assigned these tags.

```
            {
              "Sid": "AllowScopedEC2InstanceActionsWithTags",
              "Effect": "Allow",
              "Resource": [
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:fleet/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:instance/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:volume/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:network-interface/*"
              ],
              "Action": [
                "ec2:RunInstances",
                "ec2:CreateFleet"
              ],
              "Condition": {
                "StringEquals": {
                  "aws:RequestTag/kubernetes.io/cluster/${ClusterName}": "owned"
                },
                "StringLike": {
                  "aws:RequestTag/karpenter.sh/provisioner-name": "*"
                }
              }
            },

```

### AllowScopedResourceCreationTagging
The AllowScopedResourceCreationTagging Sid allows EC2 [CreateTags](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateTags.html)

> actions on `fleet`, `instance`, `volume`, `network-interface`, and `launch-template` resources, provided that a `CreateAction` of `RunInstance`, `CreateFleet`, or `CreateLaunchTemplate`

Author: Done.

                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:instance/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:volume/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:network-interface/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:launch-template/*"
              ],
              "Action": "ec2:CreateTags",
              "Condition": {
                "StringEquals": {
                  "aws:RequestTag/kubernetes.io/cluster/${ClusterName}": "owned",
                  "ec2:CreateAction": [
                    "RunInstances",
                    "CreateFleet",
                    "CreateLaunchTemplate"
                  ]
                },
                "StringLike": {
                  "aws:RequestTag/karpenter.sh/provisioner-name": "*"
                }
              }
            },
```

### AllowMachineMigrationTagging
The AllowMachineMigrationTagging Sid allows EC2 [CreateTags](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateTags.html) actions on instance resources associated with machine migrations.
These tags are set to indicate that a successful migration has occurred.

> With this action, the `kubernetes.io/cluster/${ClusterName}` tag is set to `owned` and the `karpenter.sh/managed-by` tag is set to `${ClusterName}`.

Reviewer 2: ```suggestion
This action requires that the `kubernetes.io/cluster/${ClusterName}` tag is set to `owned` to have permission to tag against the instance. If the instance has this tag, the `karpenter.sh/managed-by` tag is set to `${ClusterName}` and the `karpenter.sh/provisioner-name` tag will be set to the provisioner name value.
```

Likewise, the `karpenter.sh/provisioner-name` tag must be set to some value and any values can be set for `karpenter.sh/provisioner-name` and `karpenter.sh/managed-by`.
```
            {
              "Sid": "AllowMachineMigrationTagging",
              "Effect": "Allow",
              "Resource": "arn:${AWS::Partition}:ec2:${AWS::Region}:*:instance/*",
              "Action": "ec2:CreateTags",
              "Condition": {
                "StringEquals": {
                  "aws:ResourceTag/kubernetes.io/cluster/${ClusterName}": "owned",
                  "aws:RequestTag/karpenter.sh/managed-by": "${ClusterName}"
                },
                "StringLike": {
                  "aws:RequestTag/karpenter.sh/provisioner-name": "*"
                },
                "ForAllValues:StringEquals": {
                  "aws:TagKeys": [
                    "karpenter.sh/provisioner-name",
                    "karpenter.sh/managed-by"
                  ]
                }
              }
            },
```

### AllowScopedDeletion

> The AllowScopedDeletion Sid allows [TerminateInstances](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_TerminateInstances.html) and [DeleteLaunchTemplate](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DeleteLaunchTemplate.html) actions to delete instance and launch-template resources, provided that `karpenter.sh/provisioner-name` and `kubernetes.io/cluster/${ClusterName}` tags are set.

Reviewer 2: Add some mention that by specifying that these tags _must_ be present on the resources that we are going to delete, this ensures that Karpenter can't act against instances and launch templates that aren't directly associated with that cluster

Author: Done.

                "ec2:DescribeInstanceTypes",
                "ec2:DescribeLaunchTemplates",
                "ec2:DescribeSecurityGroups",
                "ec2:DescribeSpotPriceHistory",
                "ec2:DescribeSubnets"
              ],
              "Condition": {
                "StringEquals": {
                  "aws:RequestedRegion": "${AWS::Region}"
                }
              }
            },
```

### AllowGlobalReadActions
Because pricing information does not exist in every region at the moment, the AllowGlobalReadActions Sid allows the Karpenter controller to get product pricing information (`pricing:GetProducts`) for all related resources across all regions.
The `ssm:GetParameter" action lets the Karpenter controller read SSM parameters.

**NOTE**: If potentially sensitive information is stored in SSM parameters, you could consider restricting access to these messages further.

```
            {
              "Sid": "AllowGlobalReadActions",
              "Effect": "Allow",
              "Resource": "*",
              "Action": [
                "pricing:GetProducts",
                "ssm:GetParameter"
              ]
            },
```

> ### AllowInterruptionQueueActions

Reviewer 2: There's been no mention of an interruption queue that's been created yet, so I find this statement a little confusing. Perhaps this statement should talk about the fact that Karpenter supports interruption queues (as detailed in the interruption section of the deprovisioning conceptual docs) and that you can give Karpenter permission to access that queue by specifying the resource ARN in the CF template.

Author: I took a stab at rewording this.

For the interruption queue you created (`${KarepenterInterruptionQueue.Arn}`), the AllowInterruptionQueueActions Sid lets the Karpenter controller have permission to delete messages ([DeleteMessage](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_DeleteMessage.html)), get queue attributes ([GetQueueAttributes](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_GetQueueAttributes.html)), get queue URL ([GetQueueUrl](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_GetQueueUrl.html)), and receive messages ([ReceiveMessage](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_ReceiveMessage.html)).

```
            {
              "Sid": "AllowInterruptionQueueActions",
              "Effect": "Allow",
              "Resource": "${KarpenterInterruptionQueue.Arn}",
              "Action": [
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
                "sqs:ReceiveMessage"
              ]
            },
```

### AllowPassingInstanceRole

> The AllowPassingInstanceRole Sid gives the Karpenter controller permission to pass (`iam:PassRole`) the node role (`KarpenterNodeRole-${ClusterName}`) to the instance profile.

Reviewer 2: ```suggestion
The AllowPassingInstanceRole Sid gives the Karpenter controller permission to pass (`iam:PassRole`) the node role (`KarpenterNodeRole-${ClusterName}`) to the instance profile. This gives EC2 permission explicit permission to  use the `KarpenterNodeRole-${ClusterName}` when utilizing the instance profile [consider referencing the section above that creates the instance profile here] to launch nodes.
```

Author: Done.

> This allows EC2 to check those permissions when making an EC2 instance call, such as an EC2 launch call.

Reviewer 2: I'd drop these two sentences in favor of the clarification that I specified above.

Author: Done.

The `iam:PassedToService` restricts the permission to make these requests to only the EC2 service (`ec2.amazonaws.com`).

```
            {
              "Sid": "AllowPassingInstanceRole",
              "Effect": "Allow",
              "Resource": "arn:${AWS::Partition}:iam::${AWS::AccountId}:role/KarpenterNodeRole-${ClusterName}",
              "Action": "iam:PassRole",
              "Condition": {
                "StringEquals": {
                  "iam:PassedToService": "ec2.amazonaws.com"
                }
              }
            },
```

### AllowAPIServerEndpointDiscovery

> The Karpenter controller needs to be able to find the Kubernetes cluster's API endpoint in order to communicate with the API server.

Reviewer 2: ```suggestion
You can optionally allow the Karpenter controller to discover the Kubernetes cluster's external API endpoint to enable EC2 nodes to successfully join the EKS cluster. Note: If you are not using an EKS control plane, you will have to specify this endpoint explicitly [consider referencing the global settings docs here to the `aws.clusterEndpoint` setting.
```

Author: Done.

The AllowAPIServerEndpointDiscovery Sid allows the Karpenter controller to get that information (`eks:DescribeCluster`) for the cluster (`cluster/${ClusterName}`).
```
            {
              "Sid": "AllowAPIServerEndpointDiscovery",
              "Effect": "Allow",
              "Resource": "arn:${AWS::Partition}:eks:${AWS::Region}:${AWS::AccountId}:cluster/${ClusterName}",
              "Action": "eks:DescribeCluster"
            }
          ]
        }
```

# Interruption Handling 

> Settings in this section allow the Karpenter controller to interact with interruption queues.

Reviewer 2: ```suggestion
Settings in this section allow the Karpenter controller to stand-up an interruption queue to receive notification messages from other AWS services about the health and status of instances. For example, this interruption queue allows Karpenter to be aware of spot instance interruptions that are sent 2 minutes before spot instances are reclaimed by EC2. Adding this queue allows Karpenter to be proactive in migrating workloads to new nodes. 
```

Maybe also make reference to the interruption section in the deprovisioning documentation at the end of the paragraph

Author: Done.

> So, for example, if Spot instances are being reclaimed or a node crashes, seeing messages from these queues allows Karpenter to be proactive in moving workloads or adding new nodes.

Reviewer 2: I gave the suggestion above for how I would clarify some points here so IMO, I'd drop this sentence

Author: Done.

* AWS health events
* Spot interruptions
* Spot rebalance recommendations
* Instance state changes

The resources defined in this section include:

* KarpenterInterruptionQueue
* KarpenterInterruptionQueuePolicy
* ScheduledChangeRule
* SpotInterruptionRule
* RebalanceRule
* InstanceStateChangeRule

## KarpenterInterruptionQueue
The [AWS::SQS::Queue](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-sqs-queue.html) resource is used to create an Amazon SQS standard queue.
Properties of that resource set the `QueueName` to the name of your cluster, the time for which SQS retains each message (`MessageRetentionPeriod`) to 300 seconds, and enabling serverside-side encryption using SQS owned encryption keys (`SqsManagedSseEnabled`) to `true`.
See [SetQueueAttributes](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_SetQueueAttributes.html) for descriptions of some of these attributes.

```
  KarpenterInterruptionQueue:
    Type: AWS::SQS::Queue
    Properties:
      QueueName: !Sub "${ClusterName}"
      MessageRetentionPeriod: 300
      SqsManagedSseEnabled: true
```

## KarpenterInterruptionQueuePolicy

> The Karpenter interruption queue policy is created and applied to allow the Karpenter controller to see messages for selected services.

Reviewer 2: This queue policy actually just defines what services can push to the queue. It has nothing to do with Karpenter's ability to read from it (that's defined in the KarpenterControllerPolicy). I'd say something like "The Karpenter interruption queue policy is created to allow AWS services that we want to receive instance notifications from to push notification messages to the queue."

Author: Done.

In particular, the [AWS::SQS::QueuePolicy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-sqs-queuepolicy.html) resource here applies `EC2InterruptionPolicy` to the `KarpenterInterruptionQueue`. The policy allows [sqs:SendMessage](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_SendMessage.html) actions to `events.amazonaws.com` and `sqs.amazonaws.com` services. It also allows the `GetAtt` function to get attributes from `KarpenterInterruptionQueue.Arn`.

```
  KarpenterInterruptionQueuePolicy:
    Type: AWS::SQS::QueuePolicy
    Properties:
      Queues:
        - !Ref KarpenterInterruptionQueue
      PolicyDocument:
        Id: EC2InterruptionPolicy
        Statement:
          - Effect: Allow
            Principal:
              Service:
                - events.amazonaws.com
                - sqs.amazonaws.com
            Action: sqs:SendMessage
            Resource: !GetAtt KarpenterInterruptionQueue.Arn
```

> ## ScheduledChangeRule

Reviewer 2: I would consider creating a "rules" subsection that then defines each of these rules and what the purpose for each of these rules is.

Author: Okay. I changed the smaller rules sections into a single "Rules" section and made each on into a bullet item in that section.

> ---

Author: Good question. I think this will be used with the getting started docs. However, this is a passive doc, meaning that it is not procedural in nature. We can put it in the new Reference section you are proposing, with a pointer from the place in the Getting Started guide that uses cloudformation.yaml.

This document describes the `cloudformation.yaml` file used in that guide.
These descriptions should allow you to understand:

* What Karpenter is authorized to do with your EKS cluster and AWS resources when using the `cloudformation.yaml` file
* What permissions you need to set up if you are adding Karpenter to an existing cluster

# Review the cloudformation.yaml file

To download a particular version of `cloudformation.yaml`, set the version and use `curl` to pull the file to your local system:

```bash
export KARPENTER_VERSION={{< param "latest_release_version" >}}
curl https://raw.githubusercontent.com/aws/karpenter/"${KARPENTER_VERSION}"/website/content/en/preview/getting-started/getting-started-with-karpenter/cloudformation.yaml > cloudformation.yaml
```

Following some header information, the rest of the `cloudformation.yaml` file describes the resources that CloudFormation deploys.
The sections of that file can be grouped together under the following general headings:

> * **Node Authorization**: Creates a NodeInstanceProfile, attaches a NodeRole to it, and connects it to an IAM Identity Mapping that Karpenter uses. This defines the permissions each node managed by Karpenter has to access EC2 and other AWS resources. This doesn't actually create the IAM Identity Mapping. That part is orchestrated by `eksctl` in the Getting Started guide.

Author: Done.

> * **Karpenter Controller Authorization**:  Creates the `KarpenterControllerPolicy` that is attached to the service account.

Author: Done.

Again, the actual service account creation (`karpenter`), that is combined with the `KarpenterControllerPolicy`, is orchestrated by `eksctl` in the Getting Started guide.

> * **Interruption Handling**: The interruption handling sections of this file allow the Karpenter controller to see and respond to interruptions that occur with the nodes that Karpenter is managing. See the [Interruption](https://karpenter.sh/docs/concepts/deprovisioning/#interruption) section of the Deprovisioning page for details.

Author: Done.

> * **Interruption Handling**: The interruption handling sections of this file allow the Karpenter controller to see and respond to interruptions that occur with the nodes that Karpenter is managing. See the [Interruption](https://karpenter.sh/docs/concepts/deprovisioning/#interruption) section of the Deprovisioning page for details.

Author: Sure. I think you mean the "deprovisioning" and not "disruption" page, so I used that. Let me know if you planned to move it.

A lot of the object naming that is done by `cloudformation.yaml` is based on the following:

> * Cluster name: With a user name of `bob` the Getting Started Guide would name your cluster `bob-karpenter-demo`

Author: No. username is not a literal term or a word. We could use $USER since that is how it's set.

That name would then be appended to any name below where `${ClusterName}` is included.

> * Partition: Any time an ARN is used, it includes the partition name to identify where the object is found. In most cases, that partition name is `aws`. However, it could also be `aws-cn` (for China Regions) or `aws-us-gov` (for AWS GovCloud US Regions).

Author: Done.

# Node Authorization 

> The following sections of the `cloudformation.yaml` file set up permissions related to what Kubernetes nodes created by Karpenter can do with EC2 and other AWS features.

Author: Done.

In particular, this involves setting up an instance profile and attaching a node role to that profile with the following objects:

* KarpenterNodeRole
* KarpenterNodeInstanceProfile

## KarpenterNodeInstanceProfile
This section creates an [EC2 Instance Profile](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_use_switch-role-ec2_instance-profiles.html) that includes the node role named `KarpenterNodeRole`, with the cluster name appended.
For example, with a cluster name of `bob-karpenter-demo`, the instance profile name would look like:

`KarpenterNodeInstanceProfile-bob-karpenter-demo`

```
  KarpenterNodeInstanceProfile:
    Type: "AWS::IAM::InstanceProfile"
    Properties:
      InstanceProfileName: !Sub "KarpenterNodeInstanceProfile-${ClusterName}"
      Path: "/"
      Roles:
        - !Ref "KarpenterNodeRole"
```

> To do this manually for an existing cluster, you would find the name of the instance profile that your Kubernetes nodes are already using, and use that to associate the existing IAM role to the Karpenter-managed nodes.

Author: Done.

```bash
aws iam list-instance-profiles
```

To see information for a particular instance profile (for example, `MyRoleForInstances`), type:

```bash
aws iam get-instance-profile --instance-profile-name "MyRoleForInstances"
```
```
INSTANCEPROFILE arn:aws:iam::973227887653:instance-profile/MyRoleForInstances  2020-12-10T19:45:31+00:00       AIPA6FGHHBQS4F2FFHYYV   MyRoleForInstances     /
ROLES   arn:aws:iam::111111111111:role/MyRoleForInstances      2020-12-10T19:45:30+00:00       /       AROA6FGHHBQSQ5NKPJT2H   MyRoleForInstances
ASSUMEROLEPOLICYDOCUMENT        2012-10-17
STATEMENT       sts:AssumeRole  Allow
PRINCIPAL       ec2.amazonaws.com
```

## KarpenterNodeRole

> This section of the template  defines a node IAM role. Karpenter also needs an instance profile to associate the IAM role with EC2 instances as they launch.

Author: Done.

> This section of the template  defines a node IAM role. Karpenter also needs an instance profile to associate the IAM role with EC2 instances as they launch.

Author: Done.

        - !Sub "arn:${AWS::Partition}:iam::aws:policy/AmazonSSMManagedInstanceCore"
```

The role created here includes several AWS managed policies, which are designed to provide permissions for specific uses needed by the nodes to work with EC2 and other AWS resources. These include:

* [AmazonEKS_CNI_Policy](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonEKS_CNI_Policy.html): Provides the permissions that the Amazon VPC CNI Plugin needs to configure EKS worker nodes.
* [AmazonEKSWorkerNodePolicy](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonEKSWorkerNodePolicy.html): Lets Amazon EKS worker nodes connect to EKS Clusters.
* [AmazonEC2ContainerRegistryReadOnly](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonEC2ContainerRegistryReadOnly.html): Allows read-only access to repositories in the Amazon EC2 Container Registry.
* [AmazonSSMManagedInstanceCore](https://docs.aws.amazon.com/aws-managed-policy/latest/reference/AmazonSSMManagedInstanceCore.html): Adds AWS Systems Manager service core functions for Amazon EC2.

# Karpenter Controller Authorization 

> This section sets the permissions that the Karpenter Controller will have to create and manage EC2 and other AWS resources.

Author: Done.

> When used in the Getting Started guide, `eksctl` uses these permissions to create a service account (karpenter) that is combined with the KarpenterControllerPolicy.

Author: Done.

The resources defined in this section are associated with:

* KarpenterControllerPolicy

Because the scope of the KarpenterControllerPolicy is an AWS region, the cluster's AWS region is included in the `AllowScopedEC2InstanceActions`.

## KarpenterControllerPolicy

A `KarpenterControllerPolicy` object sets the name of the policy, then defines a set of resources and actions allowed for those resources.
For our example, the KarpenterControllerPolicy would be named: `KarpenterControllerPolicy-bob-karpenter-demo`

```
  KarpenterControllerPolicy:
    Type: AWS::IAM::ManagedPolicy
    Properties:
      ManagedPolicyName: !Sub "KarpenterControllerPolicy-${ClusterName}"
      # The PolicyDocument must be in JSON string format because we use a StringEquals condition that uses an interpolated
      # value in one of its key parameters which isn't natively supported by CloudFormation
      PolicyDocument: !Sub |
        {
          "Version": "2012-10-17",
          "Statement": [
```

> Someone wanting to add Karpenter to an existing cluster, instead of using `cloudformation.yaml`, would need to find a way to create the policy and assign that policy to the service account to use IRSA.

Author: Done.

### AllowScopedEC2InstanceActions

> The AllowScopedEC2InstanceActions statement ID (Sid) identifies a set of EC2 resources that are allowed to be used with

Author: Done.

```
            {
              "Sid": "AllowScopedEC2InstanceActions",
              "Effect": "Allow",
              "Resource": [
                "arn:${AWS::Partition}:ec2:${AWS::Region}::image/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}::snapshot/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:spot-instances-request/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:security-group/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:subnet/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:launch-template/*"
              ],
              "Action": [
                "ec2:RunInstances",
                "ec2:CreateFleet"
              ]
            },
```

### AllowScopedEC2InstanceActionsWithTags
The AllowScopedEC2InstanceActionsWithTags Sid allows the 
[RunInstances](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_RunInstances.html), [CreateFleet](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateFleet.html), and [CreateLaunchTemplate](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateLaunchTemplate.html)

> actions requested by the Karpenter controller to create all `fleet`, `instance`, `volume`, `network-interface`, or `launch-template` EC2 resources (for the partition and region), and requires that the `kubernetes.io/cluster/${ClusterName}` tag be set to `owned` and a `karpenter.sh/nodepool` tag be set to any value with these actions to ensure that Karpenter is only allowed to create instances for a single EKS cluster.

Author: Done.

> This makes sure that these resources that are managed by the Karpenter controller are assigned these tags.

Author: Done.

```
            {
              "Sid": "AllowScopedEC2InstanceActionsWithTags",
              "Effect": "Allow",
              "Resource": [
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:fleet/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:instance/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:volume/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:network-interface/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:launch-template/*"
              ],
              "Action": [
                "ec2:RunInstances",
                "ec2:CreateFleet",
                "ec2:CreateLaunchTemplate"
              ],
              "Condition": {
                "StringEquals": {
                  "aws:RequestTag/kubernetes.io/cluster/${ClusterName}": "owned"
                },
                "StringLike": {
                  "aws:RequestTag/karpenter.sh/nodepool": "*"
                }
              }
            },

```

### AllowScopedResourceCreationTagging
The AllowScopedResourceCreationTagging Sid allows EC2 [CreateTags](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateTags.html)

> actions on `fleet`, `instance`, `volume`, `network-interface`, and `launch-template` resources, While making `RunInstance`, `CreateFleet`, or `CreateLaunchTemplate` calls.

Author: Done.

```
            {
              "Sid": "AllowScopedResourceCreationTagging",
              "Effect": "Allow",
              "Resource": [
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:fleet/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:instance/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:volume/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:network-interface/*",
                "arn:${AWS::Partition}:ec2:${AWS::Region}:*:launch-template/*"
              ],
              "Action": "ec2:CreateTags",
              "Condition": {
                "StringEquals": {
                  "aws:RequestTag/kubernetes.io/cluster/${ClusterName}": "owned",
                  "ec2:CreateAction": [
                    "RunInstances",
                    "CreateFleet",
                    "CreateLaunchTemplate"
                  ]
                },
                "StringLike": {
                  "aws:RequestTag/karpenter.sh/nodepool": "*"
                }
              }
            },
```

### AllowScopedResourceTagging

> The AllowScopedResourceTagging Sid allows EC2 [CreateTags](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateTags.html) actions on instance resources associated with node migrations.

Author: Done.

These tags are set to indicate that a successful migration has occurred.
With this action, the `karpenter.sh/cluster/${ClusterName}` tag is set to `owned`.
Likewise, the `karpenter.sh/nodepool` tag must be set to some value and any values can be set for `karpenter.k8s.aws/nodeclaim`.
```
            {
              "Sid": "AllowScopedResourceTagging",
              "Effect": "Allow",
              "Resource": "arn:${AWS::Partition}:ec2:${AWS::Region}:*:instance/*",
              "Action": "ec2:CreateTags",
              "Condition": {
                "StringEquals": {
                  "aws:ResourceTag/karpenter.sh/cluster/${ClusterName}": "owned"
                },
                "StringLike": {
                  "aws:ResourceTag/karpenter.sh/nodepool": "*"
                },
                "ForAllValues:StringEquals": {
                  "aws:TagKeys": [
                    "karpenter.k8s.aws/nodeclaim",
                    "Name"
                  ]
                }
              }
            },
```

### AllowScopedDeletion

> The AllowScopedDeletion Sid allows [TerminateInstances](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_TerminateInstances.html) and [DeleteLaunchTemplate](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DeleteLaunchTemplate.html) actions to delete instance and launch-template resources, provided that `karpenter.sh/nodepool` and `kubernetes.io/cluster/${ClusterName}` tags are set.These tags must be present on all resources that Karpenter is going to delete. This ensures that Karpenter can only delete instances and launch templates that are associated with it.

Author: Done.

This allows the Karpenter controller to do any of those read-only actions across all related resources for that AWS region.

```
            {
              "Sid": "AllowRegionalReadActions",
              "Effect": "Allow",
              "Resource": "*",
              "Action": [
                "ec2:DescribeAvailabilityZones",
                "ec2:DescribeImages",
                "ec2:DescribeInstances",
                "ec2:DescribeInstanceTypeOfferings",
                "ec2:DescribeInstanceTypes",
                "ec2:DescribeLaunchTemplates",
                "ec2:DescribeSecurityGroups",
                "ec2:DescribeSpotPriceHistory",
                "ec2:DescribeSubnets"
              ],
              "Condition": {
                "StringEquals": {
                  "aws:RequestedRegion": "${AWS::Region}"
                }
              }
            },
```

### AllowSSMReadActions

> The AllowSSMReadActions Sid allows the Karpenter controller to read SSM parameters (`ssm:GetParameter`) from the current region.

Author: Done.

```
            {
              "Sid": "AllowInterruptionQueueActions",
              "Effect": "Allow",
              "Resource": "${KarpenterInterruptionQueue.Arn}",
              "Action": [
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
                "sqs:ReceiveMessage"
              ]
            },
```

### AllowPassingInstanceRole
The AllowPassingInstanceRole Sid gives the Karpenter controller permission to pass (`iam:PassRole`) the node role (`KarpenterNodeRole-${ClusterName}`) to the instance profile.

> This gives EC2 permission explicit permission to use the `KarpenterNodeRole-${ClusterName}` when utilizing the [instance profile](#karpenternodeinstanceprofile) to launch nodes.

Author: Done.

```
            {
              "Sid": "AllowPassingInstanceRole",
              "Effect": "Allow",
              "Resource": "arn:${AWS::Partition}:iam::${AWS::AccountId}:role/KarpenterNodeRole-${ClusterName}",
              "Action": "iam:PassRole",
              "Condition": {
                "StringEquals": {
                  "iam:PassedToService": "ec2.amazonaws.com"
                }
              }
            },
```

### AllowScopedInstanceProfileCreationActions
The AllowScopedInstanceProfileCreationActions Sid gives the Karpenter controller permission to create a new instance profile with [`iam:CreateInstanceProfile`](https://docs.aws.amazon.com/IAM/latest/APIReference/API_CreateInstanceProfile.html),
provided that the request is made to a cluster with `kubernetes.io/cluster/${ClusterName` set to owned and is made in the current region.

> Also, `karpenter.sh/nodeclass` must be set to some value.

Author: Done.

```
 {
              "Sid": "AllowScopedInstanceProfileCreationActions",
              "Effect": "Allow",
              "Resource": "*",
              "Action": [
                "iam:CreateInstanceProfile"
              ],
              "Condition": {
                "StringEquals": {
                  "aws:RequestTag/kubernetes.io/cluster/${ClusterName}": "owned",
                  "aws:RequestTag/topology.kubernetes.io/region": "${AWS::Region}"
                },
                "StringLike": {
                  "aws:RequestTag/karpenter.sh/nodeclass": "*"
                }
              }
            },
```

### AllowScopedInstanceProfileTagActions
The AllowScopedInstanceProfileTagActions Sid gives the Karpenter controller permission to tag an instance profile with [`iam:TagInstanceProfile`](https://docs.aws.amazon.com/IAM/latest/APIReference/API_TagInstanceProfile.html), based on the values shown below,

> Also, `karpenter.sh/nodeclass` must be set to some value.

Author: Done.

                  "aws:ResourceTag/topology.kubernetes.io/region": "${AWS::Region}",
                  "aws:RequestTag/kubernetes.io/cluster/${ClusterName}": "owned",
                  "aws:RequestTag/topology.kubernetes.io/region": "${AWS::Region}"
                },
                "StringLike": {
                  "aws:ResourceTag/karpenter.sh/nodeclass": "*",
                  "aws:RequestTag/karpenter.sh/nodeclass": "*"
                }
              }
            },
```

### AllowScopedInstanceProfileActions
The AllowScopedInstanceProfileActions Sid gives the Karpenter controller permission to perform [`iam:AddRoleToInstanceProfile`](https://docs.aws.amazon.com/IAM/latest/APIReference/API_AddRoleToInstanceProfile.html), [`iam:RemoveRoleFromInstanceProfile`](https://docs.aws.amazon.com/IAM/latest/APIReference/API_RemoveRoleFromInstanceProfile.html), and [`iam:DeleteInstanceProfile`](https://docs.aws.amazon.com/IAM/latest/APIReference/API_DeleteInstanceProfile.html) actions, 
provided that the request is made to a cluster with `kubernetes.io/cluster/${ClusterName` set to owned and is made in the current region.

> Also, `karpenter.sh/nodeclass` must be set to some value.

Author: Done.

```
            {
              "Sid": "AllowScopedInstanceProfileActions",
              "Effect": "Allow",
              "Resource": "*",
              "Action": [
                "iam:AddRoleToInstanceProfile",
                "iam:RemoveRoleFromInstanceProfile",
                "iam:DeleteInstanceProfile"
              ],
              "Condition": {
                "StringEquals": {
                  "aws:ResourceTag/kubernetes.io/cluster/${ClusterName}": "owned",
                  "aws:ResourceTag/topology.kubernetes.io/region": "${AWS::Region}"
                },
                "StringLike": {
                  "aws:ResourceTag/karpenter.sh/nodeclass": "*"
                }
              }
            },
```
### AllowInstanceProfileActions

> The AllowInstanceProfileActions Sid gives the Karpenter controller permission to perform [`iam:GetInstanceProfile`](https://docs.aws.amazon.com/IAM/latest/APIReference/API_GetInstanceProfile.html) actions to retrieve informatio about a specified instance profile.

Author: Done.

> The AllowInstanceProfileActions Sid gives the Karpenter controller permission to perform [`iam:GetInstanceProfile`](https://docs.aws.amazon.com/IAM/latest/APIReference/API_GetInstanceProfile.html) actions to retrieve informatio about a specified instance profile.

Author: Done.

> **Note**: If you are not using an EKS control plane, you will have to specify this endpoint explicitly. See the description of the `aws.clusterEndpoint` setting in the [ConfigMap](.settings/#configmap) documentation for details.

The AllowAPIServerEndpointDiscovery Sid allows the Karpenter controller to get that information (`eks:DescribeCluster`) for the cluster (`cluster/${ClusterName}`).
```
            {
              "Sid": "AllowAPIServerEndpointDiscovery",
              "Effect": "Allow",
              "Resource": "arn:${AWS::Partition}:eks:${AWS::Region}:${AWS::AccountId}:cluster/${ClusterName}",
              "Action": "eks:DescribeCluster"
            }
          ]
        }
```

# Interruption Handling 
Settings in this section allow the Karpenter controller to stand-up an interruption queue to receive notification messages from other AWS services about the health and status of instances. For example, this interruption queue allows Karpenter to be aware of spot instance interruptions that are sent 2 minutes before spot instances are reclaimed by EC2. Adding this queue allows Karpenter to be proactive in migrating workloads to new nodes.

> See the [Interruption](https://karpenter.sh/docs/concepts/deprovisioning/#interruption) section of the Deprovisioning page for details.

Reviewer 2: ```suggestion
See the [Interruption]({{< relref "disruption#interruption" >}}) section of the Deprovisioning page for details.
```

Consider using a relRef since this will always change if our domain or directories change. Also, relrefs are validated that they resolve during the hugo build step

Author: Sure. I think you mean the "deprovisioning" and not "disruption" page, so I used that. Let me know if you planned to move it.
